src/ray/rpc/gcs_server/gcs_rpc_client.h

```diff
@@ -97,8 +100,20 @@ class Executor {
                 : Status(StatusCode(reply.status().code()), reply.status().message()); \
         callback(status, reply);                                                       \
         delete executor;                                                               \
+      } else if (gcs_is_down_) {                                                       \
+        pending_requests_.emplace_back(executor);                                      \
+        if (pending_requests_.size() >                                                 \
+            ::RayConfig::instance().gcs_grpc_max_request_queued()) {                   \
+          RAY_LOG(WARNING) << "Pending queue for failed GCS request has reached the "  \
+                           << "limit. Blocking the current thread until GCS is back";  \
+          while (gcs_is_down_ && !shutdown_) {                                         \
+            CheckChannelStatus();                                                      \
+            std::this_thread::sleep_for(std::chrono::milliseconds(                     \
+                ::RayConfig::instance()                                                \
+                    .gcs_client_check_connection_status_interval_milliseconds()));     \
+          }                                                                            \
+        }                                                                              \
       } else {                                                                         \
-        gcs_service_failure_detected_(GcsServiceFailureType::RPC_DISCONNECT);          \
         executor->Retry();                                                             \
```

Comment:
For timeout, I think it's orthogonal to this issue and this one only fixes the reconnection issue and it's not handled by the current codebase as well.

> Also, do you think it is a good idea to add some grace period instead of immediately retrying here?

I think it's necessary. Internally in grpc there is retry logic (https://github.com/grpc/proposal/blob/master/A6-client-retries.md#retryable-status-codes) which we can use.

But for this one, at most, it'll cover 5 retries and the application layer should handle the timeout failure.

I think for the short term, we can just use that feature, but for long-term, we need to handle gRPC failure properly in application layer. In case of any kind of failure, application layer should be able to take care of them instead of letting the gRPC handling the logic like here (queueing and retrying).